### PR TITLE
Add GET HA-Flow Paths API endpoint

### LINKED ISSUE #5148: Add ability to read HA paths
Part of #5061

### PR DESCRIPTION
**Pull request by Fedichk**
Part of #5148

### CODE REVIEW

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java [hunk 1]
```diff
@@ -737,7 +738,12 @@ private void haFlowDeleteHub(TopologyBuilder topologyBuilder, PersistenceManager
     }
 
     private void haFlowReadBolt(TopologyBuilder topologyBuilder, PersistenceManager persistenceManager) {
-        HaFlowReadBolt bolt = new HaFlowReadBolt(persistenceManager);
+        HaFlowReadConfig config = HaFlowReadConfig.builder()
+                .readOperationRetriesLimit(topologyConfig.getYFlowReadRetriesLimit())
+                .readOperationRetryDelay(Duration.ofMillis(topologyConfig.getYFlowReadRetryDelayMillis()))
+                .lifeCycleEventComponent(ZooKeeperSpout.SPOUT_ID)
```
Comment (niksv) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java:744]: As you are using a lifeCycle you need to add `HA_FLOW_READ_BOLT` into zookeeper liveCycle. It is related to blue green deployment. 

Try to look to YFLOW_READ_BOLT in this file and do the same for HA_FLOW_READ_BOLT

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                    protectedReversePaths.put(flowPath.getHaSubFlowId(), flowPath);
+                }
+            }
+
+            for (HaSubFlow haSubFlow : haFlow.getHaSubFlows()) {
+                String haSubFlowId = haSubFlow.getHaSubFlowId();
+
+                FlowPathDtoBuilder flowPathDtoBuilder = buildFlowPathDto(haFlow, forwardPaths.get(haSubFlowId),
+                        reversePaths.get(haSubFlowId));
+                flowPathDtoBuilder.protectedPath(buildProtectedPathDto(haFlow, protectedForwardPaths.get(haSubFlowId),
```
Comment (niksv) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java:139]: Am I right that we always include protected path it the response, even if flow has no protected path?

Could you please check if we include empty protected info for common flows and Y flows? if we do not include it - please do the same

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+    private int getInPort(HaFlow haFlow, FlowPath flowPath) {
+        return FlowSideAdapter.makeIngressAdapter(haFlow, flowPath).getEndpoint().getPortNumber();
+    }
+
+    private int getOutPort(HaFlow haFlow, FlowPath flowPath) {
+        return FlowSideAdapter.makeEgressAdapter(haFlow, flowPath).getEndpoint().getPortNumber();
+    }
+
+    private boolean isProperPaths(Collection<FlowPath> paths) {
```
Comment (niksv) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java:214]: I'm not sure that it is a good name. One switch path is also proper path

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/HaFlowMapper.java [hunk 4]
```diff
@@ -98,4 +113,70 @@ public abstract class HaFlowMapper {
...
+        }
+        return null;
+    }
+
+    private List<GroupFlowPathPayload> toGroupPaths(List<FlowPathDto> paths, Predicate<FlowPathDto> predicate) {
+        return paths.stream()
+                .filter(predicate)
+                .map(pathMapper::mapGroupFlowPathPayload)
+                .collect(Collectors.toList());
+    }
```
Comment (niksv) [inline src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/HaFlowMapper.java:181]: Could you please add unit tests for the mapper?

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java (hunk 3 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java:214]: As we are checking 2 conditions, not only one switches paths, I meant to name the method like: "Are these paths good enough to be used inside the 'IntersectionComputer.calculatePathIntersectionFromSource' method". Or the shorter version which you see. If you have a better idea feel free to share it and I'll update the name of the method.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java [hunk 5]
```diff
@@ -737,7 +738,12 @@ private void haFlowDeleteHub(TopologyBuilder topologyBuilder, PersistenceManager
     }
 
     private void haFlowReadBolt(TopologyBuilder topologyBuilder, PersistenceManager persistenceManager) {
-        HaFlowReadBolt bolt = new HaFlowReadBolt(persistenceManager);
+        HaFlowReadConfig config = HaFlowReadConfig.builder()
+                .readOperationRetriesLimit(topologyConfig.getYFlowReadRetriesLimit())
```
[the remark this answers is not in the document]
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java:742]: Good catch, will rename it. Thanks!

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java [hunk 6]
```diff
@@ -737,7 +738,12 @@ private void haFlowDeleteHub(TopologyBuilder topologyBuilder, PersistenceManager
...
+                .readOperationRetryDelay(Duration.ofMillis(topologyConfig.getYFlowReadRetryDelayMillis()))
```
[the remark this answers is not in the document]
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java:743]: Good catch, will rename it. Thanks!

File: src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/HaFlowMapper.java (hunk 4 above)
Comment (Fedichk) [reply src-java/northbound-service/northbound/src/main/java/org/openkilda/northbound/converter/HaFlowMapper.java:181]: Sure, will add.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java (hunk 2 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java:139]: You are right. An empty protected path is redundant. Will fix it.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java (hunk 3 above)
Comment (niksv) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java:214]: maybe `arePathsIntersectible`

Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java:214]: Thanks for the idea. I would go with the same by mining, but a little bit simple in writing name `arePathsCrossed`

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java]: FIxed

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java]: Fixed

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java (hunk 1 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/FlowHsTopology.java:746]: Fixed

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java (hunk 2 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java]: Fixed

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java (hunk 3 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java]: Fixed if agree with the name

**Review by niksv — APPROVED** [review]

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+                        haFlowRepository.findById(haFlowId))
+                .map(haFlow -> HaFlowMapper.INSTANCE.toHaFlowDto(haFlow, flowRepository, haFlowRepository))
+                .map(HaFlowResponse::new)
+                .orElseThrow(() -> new FlowNotFoundException(haFlowId));
+    }
+
+    /**
+     * Gets HA-flow paths by ID.
+     */
+    public HaFlowPathsResponse getHaFlowPaths(@NonNull String haFlowId) throws FlowNotFoundException {
```
Comment (pkazlenka) [inline src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java:107]: This method is too long. Please, consider splitting it into smaller parts with meaningful names just for readability.

**Review by pkazlenka — APPROVED** [review]
Mostly good. Consider simplifying the code of one method.

File: src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java (hunk 7 above)
Comment (Fedichk) [reply src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java:107]: Completely agree with you. Spent a lot of time thinking about it, but didn't find a reasonable solution. 
Will try to do such refactoring during the diversity part adding.

**Review by pablomuri — APPROVED** [review]

**Pull request merged**